articles/what-is-an-abi/index.md

```diff
@@ -0,0 +1,268 @@
+
+- If the struct is less then 32 bytes in size then it will be split up into eight-byte chunks and each chunk will be given a sub-type-kind. We will determine what the sub-type-kind is based on the type-kind of the fields in it. If there are multiple fields in the eight-byte chunk then the sub-type-kind will be determined in this order Memory, Pointer, Integer, Float, Vector.
+
+Rust does this a little differently where each field is considered not each eight-byte chunk. 
+
+# Calling Conventions
+
+Now we need to think about what needs to get done when we call and return from a function. We need to (not in order):
+- Transfer control to the callee and then back to the caller
+- Save and restore the state of the caller
+- Pass data to the callee
+- Get the return value from the callee
+
+<!--
+1. Pass data to the caller
+2. Transfer control to callee
+3. Save the state of the caller
+4. Run callee code
+5. Set up the return value
+6. Restore saved state of the caller
+7. Transfer control back to the caller
+8. Finish caller code
+-->
+
+When we call a function we want our ABI defined process to be dynamic. This means that we want to be able to do all the steps correctly given that the only things that the caller will know are the signature and where the callee is defined in memory. And the only thing the callee will know is its own signature. We also want to be able to have theoretically infinite function calling depth (memory bounded <!-- better word please -->of course).
+
+So how does this get done? Well, let's look at calling the function first. 
+
+## The Stack
+
+The stack is how we will be able to keep track of who called the callee, more specifically were in the caller the callee needs to return to in memory. This is called the return address. The stack will also help with saving and resorting the caller's state so then when the callee returns to the caller, the caller can pick up where it left off.
+
+The stack works in the way you would expect from the name, i.e. we can push things on to the top of the stack and we can pop the top element from the stack. I am grossly oversimplifying the functionality of the stack but this article is not meant to talk about all of the ins and outs of the stack.
+
+### Transferring Control
+
+First a little background on how the CPU executed a list of instructions. To know which instruction is needing to be executed there is a register called the instruction pointer, `rip`, that points to the next instruction needing to be executed. This will step through the code one instruction at a time.
+
+Back to the topic, in assembly code, we would call a function with the `call <callee addr>` instruction. This will do two things; First, it will push the current value of the instruction pointer on to the stack (this is the return address) and then it will change the value of the instruction pointer to be the address of the callee function.

```

Comment:
things; First, it -> things: first, it

Reply:
stack (this is the return address) and then it  -> stack (this is the return address), and then it